src/localFunction.ts

```diff
@@ -0,0 +1,225 @@
+
+import { encodeFirestoreValue } from "./firestore/encodeFirestoreValue";
+import * as utils from "./utils";
+import { EmulatedTriggerDefinition } from "./emulator/functionsEmulatorShared";
+import { FunctionsEmulatorShell } from "./emulator/functionsEmulatorShell";
+import { AuthMode } from "./emulator/events/types";
+
+/**
+ * @class
+ * @this LocalFunction
+ * @param {object} trigger
+ * @param {object=} urls
+ * @param {object=} controller
+ */
+export default class LocalFunctionClass {
+  private url?: string;
+
+  constructor(
+    private trigger: EmulatedTriggerDefinition,
+    urls: Record<string, string>,
+    private controller: FunctionsEmulatorShell
+  ) {
+    this.url = urls[trigger.id];
+  }
+
+  private substituteParams(resource: string, params: Record<string, string>) {
+    const wildcardRegex = new RegExp("{[^/{}]*}", "g");
+    return resource.replace(wildcardRegex, (wildcard: string) => {
+      const wildcardNoBraces = wildcard.slice(1, -1); // .slice removes '{' and '}' from wildcard
+      const sub = params?.[wildcardNoBraces];
+      return sub || wildcardNoBraces + utils.randomInt(1, 9);
+    });
+  }
+  private constructCallableFunc(data: any, opts: any) {
+    opts = opts || {};
+
+    const headers: Record<string, string> = {};
+    if (opts.instanceIdToken) {
+      headers["Firebase-Instance-ID-Token"] = opts.instanceIdToken;
+    }
+
+    return request.post({
+      callback: (...args) => this.requestCallBack(...args),
+      baseUrl: this.url,
+      uri: "",
+      body: { data: data },
+      json: true,
+      headers: headers,
+    });
+  }
+
+  constructAuth(
+    authType: "USER" | "ADMIN" | "UNAUTHENTICATED",
+    auth?: AuthMode & { uid?: string; token?: object }
+  ): AuthMode {
+    if (auth?.admin || auth?.variable) {
+      return auth; // User is providing the wire auth format already.
+    }
+    if (typeof authType !== "undefined") {
+      switch (authType) {
+        case "USER":
+          return {
+            admin: false,
+            variable: {
+              uid: auth?.uid ?? "",
+              token: auth?.token ?? {},
+            },
+          };
+        case "ADMIN":
+          if (auth?.uid || auth?.token) {
+            throw new Error("authType and auth are incompatible.");
+          }
+          return { admin: true };
+        case "UNAUTHENTICATED":
+          if (auth?.uid || auth?.token) {
+            throw new Error("authType and auth are incompatible.");
+          }
+          return { admin: false };
+        default:
+          throw new Error(
+            "Unrecognized authType, valid values are: " + "ADMIN, USER, and UNAUTHENTICATED"
+          );
+      }
+    }
+    if (auth) {
+      return {
+        admin: false,
+        variable: {
+          uid: auth.uid ?? "",
+          token: auth.token || {},
+        },
+      };
+    }
+    // Default to admin
+    return { admin: true };
+  }
+
+  makeFirestoreValue(input?: any) {
```

Comment:
Should input be unknown here, since I believe its an arbitrary user provided value?